Add custom DNS settings to service definition

yongtang opened the pull request:
**\- What I did**

This fix tries to fix the issue raised in #24391 about allowing custom DNS settings to service definition.

**\- How I did it**

This fix adds `DNS`, `DNSOptions`, `DNSSearch` to service definition, as well as `--dns`, `--dns-opt`, and `--dns-search` to `service create`.

**\- How to verify it**

An integration test has been added to cover the changes in this fix.

**\- Description for the changelog**

Add `--dns`, `--dns-opt`, and `dns-search` to `service create`.

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #24391.

**NOTE: A PR in swarmkit has been created separately:**
https://github.com/docker/swarmkit/pull/1681

yongtang committed:
Add custom DNS settings to service definition

This fix tries to fix the issue raised in 24391 about allowing
custom DNS settings to service definition.

This fix adds `DNSConfig` (`Nameservers`, `Options`, `Search`) to
service definition, as well as `--dns`, `--dns-opt`, and `dns-search`
to `service create`.

An integration test has been added to cover the changes in this fix.

This fix fixes 24391.

A PR in swarmkit will be created separately.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

thaJeztah commented:
ping @aluzzardi @mrjana @dnephin PTAL

aaronlehmann commented:
Design LGTM

cpuguy83 commented:
Moving this back to design review because the UX for the flags for `create` and `update` need to be decided on.

thaJeztah commented:
Discussing this; we probably want this on `docker service update` as well, can you add that to this PR? (will be 6 flags; `--dns-add` , `--dns-rm`, etc.)

yongtang commented on api/types/swarm/container.go:
```diff
@@ -6,6 +6,19 @@ import (
+// DNSConfig specifies DNS related configurations in resolver configuration file (resolv.conf)
+// Detailed documentation is available in:
+// http://man7.org/linux/man-pages/man5/resolv.conf.5.html
+// TODO: domain is not supported yet
```
@thaJeztah `resolv.conf` allows to specify `domain` option:

http://man7.org/linux/man-pages/man5/resolv.conf.5.html
```
domain Local domain name.
```

I will update the PR to make `domain` (not supported) explicit.

yongtang committed:
Add custom DNS settings to service update

This fix adds `--dns-add`, `--dns-rm`, `--dns-opt-add`, `--dns-opt-rm`,
`--dns-search-add` and `--dns-search-rm` to `service update`.

An integration test and a unit test have been added to cover the changes in this fix.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks all for the review. The PR has been updated with DNS flags for `service update`. The docs has not been updated yet but will update the PR if we are OK with the implementation.

thaJeztah commented:
oh, needs a rebase already 😞

thaJeztah commented:
moving to code review

yongtang commented:
Thanks @thaJeztah The PR has been rebased and updated.

yongtang commented:
Ping @thaJeztah, any feedback for the PR?

mavenugo commented:
ping @mrjana @sanimej

dongluochen commented on cli/command/service/update.go:
```diff
@@ -441,6 +457,56 @@ func updateGroups(flags *pflag.FlagSet, groups *[]string) error {
+		}
+	}
+	// Sort so that result is predictable.
+	sort.Strings(newConfig.Nameservers)
```
Remove duplicate here?

dongluochen commented on cli/command/service/update_test.go:
```diff
@@ -120,6 +120,43 @@ func TestUpdateGroups(t *testing.T) {
+	flags.Set("dns-add", "2.2.2.2")
+	flags.Set("dns-rm", "3.3.3.3")
+	flags.Set("dns-add", "4.4.4.4")
+	flags.Set("dns-rm", "2.2.2.2")
```
Better to test IPv6 address and incorrect IP address.

yongtang replied:
Thanks @dongluochen. Just updated the PR with additional test cases, covering IPv6, incorrect IP address/domain name, and with duplicates. Please take a look and let me know if there are any issues.

LK4D4 commented:
@yongtang does this still needs vendoring? Also, need rebase.

yongtang commented:
Thanks @LK4D4. The PR has been rebased and the vendoring has been done already.

dongluochen commented:
@yongtang Need another rebase. 

PR LGTM (not a maintainer).

sanimej commented on docs/reference/commandline/service_update.md:
```diff
@@ -26,6 +26,12 @@ Options:
       --constraint-rm value              Remove a constraint (default [])
       --container-label-add value        Add or update container labels (default [])
       --container-label-rm value         Remove a container label by its key (default [])
+      --dns-add list                     Add or update custom DNS servers (default [])
```
DNS options of a running container can't be modified. We should provide the dns options only on service create.

sanimej commented:
@thaJeztah @yongtang  DNS configs of a running container/task can't be modified. Given that, I think we should remove the dns add/rm options for `service update`.

yongtang commented:
Thanks @sanimej. Let me take a look and remove related fields in `service update`.

thaJeztah commented:
@sanimej won't `service update` create new tasks? Or is it still an issue then?

yongtang commented:
I tried again. The service update recreated new tasks. Both `1.2.3.4 -> 8.8.8.8` and `8.8.8.8 -> 1.2.3.4` works.

sanimej commented:
@yongtang @thaJeztah If the service update creates recreates the tasks with the updated config then it shouldn't be an issue. Thanks for confirming.

thaJeztah commented:
@yongtang before I forget to ask; don't forget to document the API changes in this as well 😅

yongtang commented:
Thanks @thaJeztah. The PR has been updated with doc changes.

thaJeztah commented on docs/reference/api/docker_remote_api_v1.25.md:
```diff
@@ -5196,6 +5196,11 @@ image](#create-an-image) section for more details.
                   - **Options** - key/value map of driver specific options.
         - **StopGracePeriod** – Amount of time to wait for the container to terminate before
           forcefully killing it.
+        - **DNSConfig** – Specification for DNS related configurations in
```
Perhaps we should add it to the JSON example as well (just realized we didn't do that for the TTY option :smile:)

yongtang replied:
Updated docs for the examples as well 😅

thaJeztah commented:
ping @aaronlehmann @mrjana can you review this?

aaronlehmann commented on cli/command/service/update.go:
```diff
@@ -48,13 +48,20 @@ func newUpdateCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags.StringSliceVar(&opts.groups, flagGroupAdd, []string{}, "Add additional supplementary user groups to the container")
+	flags.Var(&opts.dns, flagDNSAdd, "Add or update custom DNS servers")
+	flags.Var(&opts.dnsOptions, flagDNSOptionsAdd, "Add or update DNS options")
+	flags.Var(&opts.dnsSearch, flagDNSSearchAdd, "Add or update custom DNS search domains")
```
Why are the remove options singular ("server") but the add options plural ("servers")?

yongtang replied:
Thanks @aaronlehmann. I updated the PR to change them into plural forms for DNS related options. The docs has been updated to reflect the changes as well.

The rest of the flags all follow the same pattern (singular for remove and plural for add). There are quite a few. I will add another PR to address the rest of the flags.

aaronlehmann commented on docs/reference/api/docker_remote_api_v1.25.md:
```diff
@@ -5196,6 +5201,11 @@ image](#create-an-image) section for more details.
+          resolver configuration file (resolv.conf).
+            - **Nameservers** – A list of the IP addresses of the name servers.
+            - **Search** – A search list for host-name lookup.
+            - **Options** – A list of internal resolver variables to be modified.
```
Specify the format for the options?

aaronlehmann commented:
Thanks for making the changes. LGTM

aluzzardi commented:
Design, Code and Docs look good.

LGTM

@yongtang can you please rebase?

stevvooe commented on cli/command/service/opts.go:
```diff
@@ -463,6 +474,15 @@ const (
+	flagDNS                   = "dns"
+	flagDNSRemove             = "dns-rm"
+	flagDNSAdd                = "dns-add"
+	flagDNSOptions            = "dns-opt"
```
Let's not abbreviate `options` here.

stevvooe commented on docs/reference/api/docker_remote_api_v1.25.md:
```diff
@@ -5459,6 +5474,11 @@ image](#create-an-image) section for more details.
                   - **Options** - key/value map of driver specific options
         - **StopGracePeriod** – Amount of time to wait for the container to terminate before
           forcefully killing it.
+        - **DNSConfig** – Specification for DNS related configurations in
```
Out of scope for this PR: why aren't we just generating this from the godoc?

thaJeztah replied:
This is gonna change once we switch to having the docs generated using swagger https://github.com/docker/docker/issues/27919

stevvooe replied:
I am not sure if swagger solves the actual problem. Can godoc integrate with swagger?

thaJeztah replied:
oh, good point @dnephin?

dnephin replied:
I don't see how godoc is relevant here. This is HTTP API documentation not go documentation.

The workflow will be "swagger spec" ->  "go code with comments" ->  "go doc".

This reference will be generated directly from the swagger spec. If we need godoc it will be available from the generated  comments.

stevvooe commented:
LGTM

thaJeztah commented:
@yongtang I talked with @stevvooe, and it looks good to him after the flag is changed to `--dns-options` (instead of `--dns-opt`)

thaJeztah commented:
@yongtang As a follow up PR, we can add an alias to `docker run` / `docker create`, like we did for `--network`. i.e., for `docker run`, the "preferred" option should be `--dns-options` as well, but we keep `--dns-opt` (possibly hidden)

yongtang commented:
Thanks @aluzzardi @stevvooe @thaJeztah. The PR has been updated with `--dns-options`.

thaJeztah reviewed (approved):
docs LGTM

vieux commented:
docs LGTM

thaJeztah commented:
Thanks so much @yongtang!

stevenao commented:
using 1.13 for windows, for some reason the container doesn't try to send query to my host's dns now. It won't resolve my private network dns.
